### PR TITLE
adding more internal random functions to torch

### PR DESCRIPTION
**Pull request by bomri**
Using other random packages, such as python Random, is not dependent on torch seed setting.
This makes it hard to maintain and reproduce experiments that use functions from these packages such as tranforms.
Even when setting seed for python Random, it is still used often in other python packages, that uses the random generator in a non consistent way which influences the stability of model training.
This PR is adding some of the popular functions, based on torch random generator which will be affected by torch seed.

### CODE REVIEW

**Comment by ezyang** [comment]
@rgommers, do you think you could help us make a call whether or not this should go in PyTorch, based on your experience in Numpy?

**Comment by rgommers** [comment]
@ezyang yes, I'll have a look

**Review by rgommers — COMMENTED** [review]
The rationale given here for expanding the random number generation functionality probably makes sense. There's currently a limited amount of distributions provided, and for programs it's quite important to be able to use a _single_ pseudorandom_ generator instance in the whole program. That means there's a stronger reason to put `random` inside PyTorch itself than, e.g., `linalg` and other functionality that could be done separately.

That said, this PR doesn't look right. Some issues:
- Functions added are fairly arbitrary and are all one-liners
- `choice` and `sample` are good function to have in principle (not in this form though), while `random_float`, `uniform` and `random_int` don't look useful.
- When choosing functions that are already in NumPy, the APIs should be made to match unless there's a good reason not to. Taking `choice` for example, compare `torch.random.choice(input)` with `numpy.random.choice(a, size=None, replace=True, p=None)`.

The PyTorch RNG support is slightly incoherent at the moment as far as I can tell. 
- `torch.Generator` has some methods; `torch.random` a few more functions that correspond to those methods, plus a couple more (`default_generator`, `fork_rng`).
- Distributions like `bernoulli`, `cauchy` and `log_normal` are `Tensor` methods rather than `Generator` methods. That probably has a reason (device handling?), but it looks a little odd and will be messy when one wants to add more distributions.
- Distributions aren't treated the same, e.g. there's `torch.bernoulli` but no `torch.cauchy`.
- Why is there a `torch.random` namespace at all if `torch.rand` et al. live in the main namespace?
- The [random sampling](https://pytorch.org/docs/stable/torch.html#random-sampling) functions and the [in-place variants](https://pytorch.org/docs/stable/torch.html#in-place-random-sampling) don't match 1:1 (also, I'd have preferred an `inplace` keyword or some other way of not doubling the number of functions/methods). 

For context: the separate `np.random.<rng_func>` functions are considered a mistake by the NumPy devs. The reason is that they're relying on global state, and it's very much not robust in multi-threaded applications (plus some other reasons). Those functions are also methods on the `RandomState` generator, which is the better practice. On top of that, all that stuff is now legacy and replaced with a completely new, pluggable infrastructure and higher quality generators that explicitly support parallel usage. That infrastructure could perhaps be reused/adapted.

tl;dr I would reject this PR, but be +1 on writing a proposal for if and how RNG functionality should be expanded or changed.

**Comment by ezyang** [comment]
Thanks! I agree with the review. Also cc'ing @pbelevich and @nairbv who are working in the area of rngs in PyTorch
